fix: inventory desync when interaction is cancelled

Emilxyz opened the pull request:
Fixes to a9399451481e3f8b376484e63909ff6c274f03db that caused inventory desyncs when the client predicted an interaction would result in an item being added to its inventory

Fixes #12947 Fixes #12954 Fixes #13253

Emilxyz committed:
fix: inventory desync when interaction is cancelled

Emilxyz commented:
Okay so I've added a `clientPredictsInteraction` method to keep track of which block & item interactions actually need this. I couldn't really find a good place to put it, so I just slapped it into MCUtil for now. Maybe this should go in its own util class, something like `ClientPredictions`? idk

I scrolled through the creative inventory and tested some blocks that looked like they might be affected by this, but I didn't find anything else. So if anyone knows of any I missed, please let me know.

I added a check to this method on both interact & spawn protection to send the correct update, but I'm a bit confused by
> Only call this for the spawn protection cases

Isn't that what's already being done? Should I split this if check up into something like
```java
if (awaitingPositionFromClient == null) {
  if (!mayInteract && !...) {
    send correct update
    return;
  }
  rest of the code
}
```
<img width="1298" height="645" alt="image" src="https://github.com/user-attachments/assets/fa85b858-fdd1-4686-8a97-8b456b819980" />

Emilxyz commented:
Implemented the spawn prot thing.

From my testing and understanding you can only place a candle on a cake and not remove it, unless you eat or destroy the cake. I did find something else tho: picking up fluids or bee hives using a bucket or glass bottle with a count > 1. 
I found no good way to get the fluid they're interacting with that early in the interact, other than using Level#clip. but calling that an extra time per interaction just for this seemed unnecessary. So I decided to just check if they're holding a bucket/glass bottle with a count > 1. That conveniently also covers the bee nest.

<details>
<summary>Screenshot from my test world, with other stuff I tested</summary>
<img width="1916" height="1055" alt="image" src="https://github.com/user-attachments/assets/20a35938-53bb-4d20-90d0-035f492e55dc" />
</details>

Lulu13022002 commented:
You can't really rely on tags, since datapacks may remove/change their content and you still need to resend held item within the spawn protection when clientPredictsInteraction = false

Lulu13022002 reviewed (commented):
Seems cancelling the PlayerBucketEntityEvent in creative (or with some custom water bucket with a max stack size >1) doesn't update the right slot too so make sure to test all the other place.

Emilxyz commented:
Maybe this time I won't think of 5 more things the instant I click ready for review. More changes:
- Added a case for cancelling converting mud, either though PlayerInteractEvent or EntityChangeBlockEvent
- Added a separate clientPredictsInteraction method for entity interactions
    - Milking AbstractCow
    - Getting soup from MushroomCow
    - Bucketing Bucketables
- Refactored the clientPredictsInteraction method into multiple if statements, instead of just one big one, to improve readability

Lulu13022002 commented:
Can you look at the InventoryDragEvent, i believe the above commit broke the ability for plugins to set the cursor on the event when the carried item was empty (which might happens at the end of the action when all the items have been placed).

Specifically, this comment is not really accurate: https://github.com/PaperMC/Paper/blob/main/paper-server/patches/sources/net/minecraft/world/inventory/AbstractContainerMenu.java.patch#L152

Emilxyz committed:
fix: always set carried item after successful InventoryDragEvent

Emilxyz commented:
You were right, this broke setting the carried item when it's empty. I'm not sure what behavior was even fixed with that if statement, I did some testing by just removing it and everything seemed to work as expected when the inventory is closed server side. To me it doesn't look like getCarried can return null anyways so this is basically what was done before the commit.

https://github.com/user-attachments/assets/c7e34756-f29a-4aa3-bca5-62dfcd18d8b4

In the first test I kept holding mouse one not letting the event fire and it dropped the old item stack. The second test I release mouse 1 and the event changes my carried item to a diamond sword and that is dropped when the inventory is closed. Am I missing something obvious here?

Emilxyz committed:
fix: forceHeldSlot when setting PlayerItemConsumeEvent replacement

Emilxyz committed:
fix: inventory desync with bucket event result item

kennytv committed:
Partially address reviews

kennytv reviewed (approved):
The empty/fill events can be addressed later
